src/tasks/transaction_sender.rs

```diff
@@ -1,188 +1,221 @@
+use tracing::{instrument, trace, warn};
 
 use crate::{
     client::SEND_TRANSACTION_INTERVAL,
-    messages::{BlockMessage, ConfirmTransactionMessage, SendTransactionMessage},
+    messages::{BlockMessage, ConfirmTransactionMessage, SendTransactionMessage, StatusMessage},
 };
 
-/// Spawns an independent task that listens for [`SendTransactionMessage`]s and periodically submits
-/// transactions using the Solana RPC client, re-signing the transactions when necessary.
-///
-/// It does *not* check the outcome of the transaction at all other than failing if the transaction
-/// submission itself fails. When this happens, the transaction will be queued for re-sending.
-///
-/// The task will exit if there are no transaction senders alive. This will happen when the
-/// [BatchClient](`crate::batch_client::BatchClient`) has been dropped.
-#[allow(clippy::too_many_arguments)]
-pub fn spawn_transaction_sender(
+#[derive(Debug, thiserror::Error)]
+pub enum SenderError {
+    #[error("RPC error: {0}")]
+    RpcError(#[from] Box<Error>),
+    #[error("confirm channel closed")]
+    ConfirmChannelClosed,
+    #[error("send channel closed")]
+    SendChannelClosed,
+    #[error("response channel closed")]
+    ResponseChannelClosed,
+}
+
+pub type SenderResult<T = ()> = Result<T, SenderError>;
+
+impl SenderError {
+    pub fn is_transient(&self) -> bool {
+        match self {
+            SenderError::RpcError(e) => match e.kind() {
+                ClientErrorKind::SigningError(_) => true,
+                ClientErrorKind::TransactionError(transaction_error) => !matches!(
+                    transaction_error,
+                    TransactionError::AccountInUse
+                        | TransactionError::AccountLoadedTwice
+                        | TransactionError::AccountNotFound
+                        | TransactionError::ProgramAccountNotFound
+                        | TransactionError::InsufficientFundsForFee
+                        | TransactionError::AlreadyProcessed
+                        | TransactionError::CallChainTooDeep
+                        | TransactionError::InvalidAccountIndex
+                        | TransactionError::InvalidProgramForExecution
+                        | TransactionError::AccountBorrowOutstanding
+                        | TransactionError::UnsupportedVersion
+                        | TransactionError::InvalidWritableAccount
+                        | TransactionError::TooManyAccountLocks
+                        | TransactionError::AddressLookupTableNotFound
+                        | TransactionError::InvalidAddressLookupTableOwner
+                        | TransactionError::InvalidAddressLookupTableData
+                        | TransactionError::InvalidAddressLookupTableIndex
+                        | TransactionError::InvalidRentPayingAccount
+                        | TransactionError::InsufficientFundsForRent { .. }
+                        | TransactionError::UnbalancedTransaction
```

Comment:
Why not match against the errors that are actually transient instead of the ones that aren't?

Reply:
Updated the implementation